### PR TITLE
Implemented Voxel volumes and improved performance

### PR DESCRIPTION
**Pull request by john-sharratt**
- Fixed a major bug where the `le` function in `TUVec` was implemented incorrectly causing the `contains` function to return incorrect values
- Implemented a new `Volume` interface without breaking the existing `Position` interface
- Removed the `map` `Pool` which is no longer needed as the remove performance is now natively much faster and besides this map does not work with `Volumes` that can have `ElementId` values in multiple regions of the octree.
- Optimized the performance or `insert`, `remove` and other operations (see benchmark)

Volume volumes allow inserting a voxel that holds a specific region of space into the octree ensuring that if it overlaps existing voxels that it merges around them. This capability is especially important for ray tracing and sphere tracing needed for collision detection.

### CODE REVIEW

**Comment by john-sharratt** [comment]
image

File: src/tree.rs [hunk 1]
```diff
@@ -241,24 +264,126 @@ where
...
+                        if volume.min.z <= branch_center.z {
+                            removals.push(Removal {
+                                parent: Some(node),
+                                node: branch.x0_y0_z0(),
+                            });
+                        }
+                        if volume.max.z >= branch_center.z {
+                            removals.push(Removal {
+                                parent: Some(node),
+                                node: branch.x0_y0_z1(),
```
Comment (exor2008) [inline src/tree.rs]: Can we put that code repeating in `fn _insert` in a separate function and mark it `inline`?

**Review by exor2008 — COMMENTED** [review]
Repeating code

**Comment by exor2008** [comment]
So all those counting of the filled children and caching `NodeId`s in `map` didn't help a lot, huh?)
Surprising that naive removing with constant searching could be faster then remembering `NodeId` and just looking for it by an offset in the `map`.
Early optimizations... =)

**Comment by exor2008** [comment]
Don't you mind to fix those clippy warnings?

File: src/node.rs [hunk 2]
```diff
@@ -93,25 +80,49 @@ impl Branch {
         }
     }
 
-    pub(crate) fn from_filled(children: [NodeId; 8], filled: u8) -> Self {
-        Branch { children, filled }
+    pub fn x0_y0_z0(&self) -> NodeId {
```
Comment (exor2008) [inline src/node.rs:81]: Does it make sense to mark these 8 small functions with `#[inline]`?

**Review by exor2008 — COMMENTED** [review]
Make it inline?

**Comment by john-sharratt** [comment]
All good, I'll fix the clippy errors and your others comments in a bit.

Indeed the map lookup and the `filled` counts were just not making much difference in the end and were keeping the complexity too high when trying to do volume voxels - probably because of cache line coherency benefits in modern CPU's as it walks the tree. The remove function is now much faster while the others did not regress.

File: src/node.rs (hunk 2 above)
Comment (john-sharratt) [reply src/node.rs:81]: Done

File: src/tree.rs (hunk 1 above)
Comment (john-sharratt) [reply src/tree.rs]: Done

**Comment by john-sharratt** [comment]
All done

**Pull request merged**